File: ecosystem/plot_images_types.py
#!/usr/bin/env python

# Plots the commonly used base image types.

import pandas as pd
import matplotlib.pyplot as plt

# read CSV with base image count:
df_images_count = pd.read_csv('./data/images_count.csv')
df_images_count_top100 = pd.read_csv('./data/images_count_top100.csv')
df_images_count_top1000 = pd.read_csv('./data/images_count_top1000.csv')


print ''
print '####################################'
print '# Number of base images used'
print '####################################'
count_overall = df_images_count['count'].sum()
count_top100 = df_images_count_top100['count'].sum()
count_top1000 = df_images_count_top1000['count'].sum()
print 'Overall: %s' % count_overall
print 'Top 1000: %s' % count_top1000
print 'Top 100: %s' % count_top100


print ''
print '####################################'
print '# Number of distinct images'
print '####################################'
print df_images_count['imagename'].count()


df_top_images = df_images_count.sort_values('count', ascending=False)[:25].set_index('imagename')
df_top_images['percent'] = 100 * df_top_images['count'] / df_top_images['count'].sum()
# print df_top_images

df_top_images_top1000 = df_images_count_top1000.sort_values('count', ascending=False)[:25].set_index('imagename')
print df_top_images_top1000
df_top_images_top1000['percent 1000'] = 100 * df_top_images_top1000['count'] / df_top_images_top1000['count'].sum()

df_top_images_top100 = df_images_count_top100.sort_values('count', ascending=False)[:25].set_index('imagename')
df_top_images_top100['percent 100'] = 100 * df_top_images_top100['count'] / df_top_images_top100['count'].sum()

df_top_images_types = pd.read_csv('./data/images_top25_types.csv').set_index('imagename')

df_top_combined = pd.concat([df_top_images['percent'], df_top_images_top1000['percent 1000'], df_top_images_top100['percent 100'], df_top_images_types], axis=1)
df = pd.concat([
  df_top_combined.groupby('type')['percent'].sum(),
  df_top_combined.groupby('type')['percent 1000'].sum(),
  df_top_combined.groupby('type')['percent 100'].sum()
], axis=1)
df.columns = ['Overall', 'Top 1000', 'Top 100']
df = df.sort_values('Overall', ascending=False)

# plot it:
ax = df.plot(kind='bar', rot=0)
ax.legend(labels=['All', 'Top-1000', 'Top-100'])
ax.set_ylabel('Base image types [%]')
ax.set_xlabel('')

fig = ax.get_figure()
fig.set_figwidth(8)
fig.set_figheight(4)
fig.tight_layout()
fig.savefig('fig/top_images_types.png')

File: ecosystem/plot_images_types.R
library(ggplot2) 

# Plots the commonly used base image types

# home <- "~/git_repos/papers/msr17-docker-evolution/analyses/ecosystem/"
home <- ""

image_types = read.csv(paste(home, "data/images_top25_types.csv", sep = ""))
images = read.csv(paste(home, "data/images_count.csv", sep = ""))
images <- merge(images, image_types, by="imagename")
images = images[order(-1 * images$count),][0:25,]
images_top100 = read.csv(paste(home, "data/images_count_top100.csv", sep = ""))
images_top100 <- merge(images_top100, image_types, by="imagename")
images_top100 = images_top100[order(-1 * images_top100$count),][0:25,]
images_top1000 = read.csv(paste(home, "data/images_count_top1000.csv", sep = ""))
images_top1000 <- merge(images_top1000, image_types, by="imagename")
images_top1000 = images_top1000[order(-1 * images_top1000$count),][0:25,]

types <- c("OS", "Language runtime", "Application", "Other", "Variable")

df <- data.frame("Type"=integer(),
                 "Projects"=double(),
                 "DataSet" = character(), 
                 stringsAsFactors=FALSE) 

for(type in types) {
  count_all <- 100 * sum(images[images$type == type,]$count) / sum(images$count)
  count_top100 <- 100 * sum(images_top100[images_top100$type == type,]$count) / sum(images_top100$count)
  count_top1000 <- 100 * sum(images_top1000[images_top1000$type == type,]$count) / sum(images_top1000$count)
  df[nrow(df) + 1, ] <- list(type, count_all, "All")
  df[nrow(df) + 1, ] <- list(type, count_top100, "Top-100")
  df[nrow(df) + 1, ] <- list(type, count_top1000, "Top-1000")
}
df$DataSet <- factor(df$DataSet)
df$Type <- factor(df$Type, levels=types)

cbPalette <- c("#3333FF", "#208222", "#C43428", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#D55E00")
ggplot(df, aes(x=Type, y=Projects, fill=DataSet)) +  geom_bar(stat="identity", position=position_dodge()) + ylab("% of Projects with Base Image of Type") + theme_bw() + theme(legend.justification=c(1,1), legend.position=c(1,1), legend.background = element_rect(colour = "black"), legend.title = element_blank()) + scale_fill_manual(values=cbPalette) + scale_y_continuous(
  breaks = c(0, 10, 20, 30, 40, 50, 60),
  labels = c(0, 10, 20, 30, 40, 50, 60)
)
ggsave("fig/top_image_types_r.pdf", width = 6.0, height = 3.56, units="in")
